File: ctrl.c
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/wait.h>

int h, fd;

void timeout (int s)
{
    kill(fd, SIGKILL);
}

int main(int argc, char** argv)
{
    h = atoi(argv[1]);

    while (1){

        fd = fork();
        if (fd == 0){
            char str[2];
            sprintf(str, "%d", h);
            
            int out = open("pddl.out", O_RDWR|O_CREAT|O_TRUNC, 0666);
            close(1);
            dup(out);
            execlp("./ff", "./ff", "-O",  "-o", "domain.pddl", "-f", "problem.pddl", "-h", str, (char * ) NULL);
        } else {
            struct sigaction sa;
            sa.sa_handler = &timeout;
            sa.sa_flags = SA_RESTART;
            sigfillset(&sa.sa_mask);
            sigaction(SIGALRM, &sa, NULL);

            alarm(6);
            
            int ret, pid;

            while ((pid = waitpid(-1, &ret, 0) > 0));


            if (WIFEXITED(ret))
            {
                exit(0);
            }
            else ++h;

        }

    }
}

File: ctrl.py
import os
import signal
import sys
import time

h = int(input("Enter the value of h: "))

while True:
    pid = os.fork()
    if pid == 0:
        str_h = str(h)
        outfile = open("pddl.out", "w")
        os.dup2(outfile.fileno(), sys.stdout.fileno())
        os.execvp("./ff", ["./ff", "-O", "-o", "domain.pddl", "-f", "problem.pddl", "-h", str_h])
    else:
        signal.signal(signal.SIGALRM, signal.SIG_IGN)
        signal.alarm(6)
        pid, ret = os.waitpid(-1, 0)
        print("Alarm here!")
        signal.alarm(0)
        if os.WIFEXITED(ret):
            break
        else:
            h += 1
